tools/autograd/gen_variable_type.py

```diff
@@ -1189,7 +1189,7 @@ def emit_check_inplace():
     def emit_increment_version():
         if not modifies_arguments:
             return []
-        return ['increment_version({});'.format(arg['name']) for arg in differentiable_outputs]
+        return ['increment_version({});'.format(arg['name']) for arg in returns]
```

Comment:
Why is this changed?  does it actually matter?
Do we have inplace ops that return multiple things? And if so do we have some that return a mix of differentiable/non-differentiable outputs?

Reply:
We should increment version on all tensors that get mutated, not just the differentiable ones. You can save non differentiable tensors as part of backwards formula...

Reply:
But I don't expect that these other outputs are actually modified inplace! We would be bumping the version of a Tensor we don't change inplace.

Also this is BC-breaking right?

Reply:
> Why is this changed? does it actually matter?
> Do we have inplace ops that return multiple things? And if so do we have some that return a mix of differentiable/non-differentiable outputs?

There are cases where it returns multiple things mixed of differentiable/non-differentiable outputs, e.g.:
https://github.com/ljk53/pytorch/commit/9ab9e4bc3489c9b3391bed174305d7a6b00457b5#diff-79b1a31c97eee8dda9e0dae02162beecR2986

non-differentiable outputs are usually things like `indices`.

Reply:
Ho right these `out=(val, ind)` functions...
Ok!

Reply:
> But I don't expect that these other outputs are actually modified inplace! We would be bumping the version of a Tensor we don't change inplace.
> 
> Also this is BC-breaking right?

This is a good point - just to clarify: regarding BC breakage, do you mean:
1. code that uses `x._version` to read variable version explicitly, or
2. code broken by falsely incrementing version for tensors that are not actually updated?

I call out "falsely" bumping because it's harmful as the breakage is totally unnecessary compared to truly bumping - in which case it is technically more correct behavior - if truly bumping breaks any code then it probably reveals bugs.

Seems we made effort to differentiate `Tensor&` and `const Tensor&` and seems `returns` at this place only contains `Tensor&` ones - so I assumed that these params are possibly mutated and bumped their versions. By eyeballing check seems in most cases the `returns` are indeed mutated, but I haven't verified all cases - do you know whether is it right assumption?

Reply:
> code that uses x._version to read variable version explicitly

This one is ok to break as this is an internal API. You might need to update a couple tests but it should be fine overall (even though we want to avoid breaking it often).

> code broken by falsely incrementing version for tensors that are not actually updated?

There are two cases here indeed:
- Cases where we were not incrementing while we should:
```python
import torch
from torch.utils import checkpoint

a = torch.ones(10, 10, requires_grad=True)

b, ind = a.max(dim=0)

with torch.no_grad():
    if False:
        ind += 1 # Raise an error as expected
    elif True:
        # No error and wrong grad
        t = torch.zeros(10)
        t[2] = 1
        torch.cummax(t, dim=0, out=(torch.Tensor(), ind))
    else:
        pass

b.sum().backward()
print(a.grad)
```
- Cases where we should not increment as the second result is not modified inplace. Not sure if this happens in practice. In any case, we should be able to tell from the signature if it is modified or not (the signature is (should) always right!).